### PR TITLE
[SPARK-47345][SQL][TESTS] Xml functions suite

### PR DESCRIPTION
**Pull request by yhosny**
### What changes were proposed in this pull request?

Convert JsonFunctiosnSuite.scala to XML equivalent. Note that XML doesn’t implement all json functions like json_tuple, get_json_object, etc.

### Why are the changes needed?

Improve unit test coverage.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by zhengruifeng** [comment]
add `[TESTS]` since it only adds new tests

**Review by MaxGekk — COMMENTED** [review]
It seems the test failure [Run / Build modules: pyspark-connect](https://github.com/yhosny/spark/actions/runs/8239241065/job/22532077826#logs) is not related to the new test suite.

File: sql/core/src/test/scala/org/apache/spark/sql/XmlFunctionsSuite.scala [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+        ArrayType(LongType, true), true) :: Nil),
+      true) :: Nil)
+
+    assert(out.schema == expected)
+  }
+
+  test("infers schemas using options") {
+    val df = spark.range(1)
+      .select(schema_of_xml(lit("<ROW><a>1</a></ROW>"),
+              Map("allowUnquotedFieldNames" -> "true").asJava))
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/XmlFunctionsSuite.scala:368]: please, fix indentation here.

File: sql/core/src/test/scala/org/apache/spark/sql/XmlFunctionsSuite.scala [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types._
+
+class XmlFunctionsSuite extends QueryTest with SharedSparkSession {
+  import testImplicits._
+
+
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/XmlFunctionsSuite.scala:33]: remove this line

**Review by MaxGekk — APPROVED** [review]
LGTM except of minor comments.

**Comment by MaxGekk** [comment]
I have checked the new test suite locally:
```
$ build/sbt "sql/test:testOnly *XmlFunctionsSuite"
[info] XmlFunctionsSuite:
22:50:44.233 WARN org.apache.hadoop.util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
[info] - from_xml (5 seconds, 317 milliseconds)
[info] - from_xml with option (timestampFormat) (382 milliseconds)
[info] - from_xml with option (rowTag) (153 milliseconds)
[info] - from_xml with option (dateFormat) (180 milliseconds)
...
[info] All tests passed.
```

+1, LGTM. Merging to master.
Thank you, @yhosny and @zhengruifeng for review.

File: sql/core/src/test/scala/org/apache/spark/sql/XmlFunctionsSuite.scala [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+        Seq(("""<ROW><i>1</i></ROW>""", "i int")).toDF("xml", "schema")
+          .select(from_xml($"xml", $"schema", options)).collect()
+      },
+      errorClass = "INVALID_SCHEMA.NON_STRING_LITERAL",
+      parameters = Map("inputSchema" -> "\"schema\""),
+      context = ExpectedContext(fragment = "from_xml", getCurrentClassCallSitePattern)
+    )
+  }
+
+  test("schema_of_xml - infers the schema of foldable JSON string") {
```
Comment (sandip-db) [inline sql/core/src/test/scala/org/apache/spark/sql/XmlFunctionsSuite.scala:453]: JSON -> XML

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/XmlFunctionsSuite.scala:453]: Made a followup: https://github.com/apache/spark/pull/45577

**Pull request closed without merging**